### ISSUE TITLE
cls parameter of classmethod raises error when bound to Callable

### ISSUE THREAD

**Issue by ringohoffman** [microsoft/pyright#4447]
**Describe the bug**
When binding the `cls` parameter of a `classmethod` to a `Callable`, e.g. for defining a special alternate constructor, an error is raised stating that `Type of parameter "cls" must be a supertype of its class "Type[...]"`.

**To Reproduce**
```python
from __future__ import annotations

from typing import Callable, Concatenate, ParamSpec

P = ParamSpec("P")


class A:
    def __init__(self, a_param_1: str) -> None:
        ...

    @classmethod
    def add_params(
        cls: Callable[P, A],  # PylancereportGeneralTypeIssues: Type of parameter "cls" must be a supertype of its class "Type[A]"
    ) -> Callable[Concatenate[float, P], A]:
        def new_constructor(i: float, *args: P.args, **kwargs: P.kwargs) -> A:
            return cls(*args, **kwargs)
        return new_constructor
    
    @classmethod
    def remove_params(
        cls: Callable[Concatenate[str, P], A],  # PylancereportGeneralTypeIssues: Type of parameter "cls" must be a supertype of its class "Type[A]"
    ) -> Callable[P, A]:
        def new_constructor(*args: P.args, **kwargs: P.kwargs) -> A:
            return cls("my_special_str", *args, **kwargs)
        return new_constructor


A.add_params()(1.0, "my_str")
A.remove_params()()

```

**Expected behavior**
I think binding `cls` to a `Callable` should be allowed.

**VS Code extension v2023.1.11**

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.289, which I just published. It will also be included in a future release of pylance.

**Comment by ringohoffman** [comment]
There was a regression in support for this in 1.1.335.

```console
$ pyright a.py
  a.py:12:6 - error: Argument of type "(cls: (**P@add_params) -> A) -> ((float, **P@add_params) -> A)" cannot be assigned to parameter "__f" of type "(type[_T@classmethod], **_P@classmethod) -> _R_co@classmethod" in function "__init__"
    Type "(cls: (**P@add_params) -> A) -> ((float, **P@add_params) -> A)" cannot be assigned to type "(type[_T@classmethod], **_P@classmethod) -> _R_co@classmethod"
      Parameter 1: type "type[_T@classmethod]" cannot be assigned to type "(**P@add_params) -> A"
        Type "type[_T@classmethod]" cannot be assigned to type "(**P@add_params) -> A"
          Function return type "object*" is incompatible with type "A"
            "object*" is incompatible with "A" (reportGeneralTypeIssues)
  a.py:20:6 - error: Argument of type "(cls: (str, **P@remove_params) -> A) -> ((**P@remove_params) -> A)" cannot be assigned to parameter "__f" of type "(type[_T@classmethod], **_P@classmethod) -> _R_co@classmethod" in function "__init__"
    Type "(cls: (str, **P@remove_params) -> A) -> ((**P@remove_params) -> A)" cannot be assigned to type "(type[_T@classmethod], **_P@classmethod) -> _R_co@classmethod"
      Parameter 1: type "type[_T@classmethod]" cannot be assigned to type "(str, **P@remove_params) -> A"
        Type "type[_T@classmethod]" cannot be assigned to type "(str, **P@remove_params) -> A"
          Function return type "object*" is incompatible with type "A"
            "object*" is incompatible with "A" (reportGeneralTypeIssues)
```
